File: DataStructure/Queue/207_canFinish.java
package LeetCode.Binary_Tree;

import java.util.Queue;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

// 到底该用什么数据类型？？
public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegrees = new int[numCourses];
        ArrayList<ArrayList<Integer>> adjacency = new ArrayList<>(); //
        // List<List<Integer>> adjacency = new ArrayList<>();
        Queue<Integer> queue = new LinkedList<>();

        for (int i = 0; i < numCourses; i++) 
            adjacency.add(new ArrayList<>());  
        
        for (int[] cp : prerequisites) {
            indegrees[cp[0]]++;
            adjacency.get(cp[1]).add(cp[0]);
        }

        for (int i = 0; i < numCourses; i++) 
            if (indegrees[i] == 0)
                queue.add(i);

        while (!queue.isEmpty()) {
            int pre = queue.poll();
            numCourses--;
            for (int cur : adjacency.get(pre))
                if (--indegrees[cur] == 0)
                    queue.add(cur);
        }
        return numCourses == 0;
    }
}

File: DataStructure/Queue/207_canFinish.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        int n = prerequisites.size();
        
        // 建立入度表
        for (int i = 0; i < n; i++) {
            indegree[prerequisites[i][0]]++;
        }
        // 建立图
        vector<vector<int>> graph(numCourses, vector<int>(numCourses));
        for (int i = 0; i < n; i++) {
            graph[prerequisites[i][1]][prerequisites[i][0]] = 1;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0)
                q.push(i);
        }
        int cnt = numCourses;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            cnt--;
            for (int i = 0; i < numCourses; i++) {
                if (graph[node][i] == 1) {
                    indegree[i]--;
                    if (indegree[i] == 0)
                        q.push(i);
                }
            }
        }
        return cnt == 0;
    }
};
// 通过了

int main() {
    Solution a;
    vector<vector<int>> pre;
    vector<int> v(2);
    v[0] = 1;
    v[1] = 0;
    pre.push_back(v);
    v[0] = 2;
    v[1] = 1;
    pre.push_back(v);
    a.canFinish(3, pre);
}
